orttraining/orttraining/training_ops/cpu/optimizer/adamw/adamw.cc

```diff
@@ -0,0 +1,224 @@
+    Tensor& weight = const_cast<Tensor&>(p.weights->Get(weight_index));
+    Tensor& gradient = const_cast<Tensor&>(p.gradients->Get(weight_index));
+    Tensor& momentums_1 = const_cast<Tensor&>(p.momentums_1->Get(weight_index));
+    Tensor& momentums_2 = const_cast<Tensor&>(p.momentums_2->Get(weight_index));
+
+    // Perform weight decay.
+    MakeEigenArrayMap<T>(weight) = MakeEigenArrayMap<T>(weight) - (MakeEigenArrayMap<T>(weight) * lr * weight_decay_);
+
+    // Compute exponentially-averaged historical gradient.
+    MakeEigenArrayMap<T>(momentums_1) = alpha_ * MakeEigenArrayMap<T>(momentums_1) +
+                                        (1.f - alpha_) * MakeEigenArrayMap<T>(gradient);
+
+    // Compute exponentially-averaged historical squared gradient.
+    MakeEigenArrayMap<T>(momentums_2) = beta_ * MakeEigenArrayMap<T>(momentums_2) +
+                                        (1.f - beta_) * MakeEigenArrayMap<T>(gradient) * MakeEigenArrayMap<T>(gradient);
+
+    // Compute the new weight.
+    auto denom = (MakeEigenArrayMap<T>(momentums_2) / beta_correction).sqrt() + epsilon_;
+    MakeEigenArrayMap<T>(weight) = MakeEigenArrayMap<T>(weight) -
+                                   (lr * MakeEigenArrayMap<T>(momentums_1)) / (alpha_correction * denom);
+  }
+
+  return Status::OK();
+}
+
+template <typename T>
+Status AdamWOptimizer<T>::AdamWComputeMode1(AdamWOptimizerBase::Prepare& p, float lr, float lr_corrected) const {
+  for (size_t weight_index = 0; weight_index < p.num_of_weights; ++weight_index) {
+    Tensor& weight = const_cast<Tensor&>(p.weights->Get(weight_index));
+    Tensor& gradient = const_cast<Tensor&>(p.gradients->Get(weight_index));
+    Tensor& momentums_1 = const_cast<Tensor&>(p.momentums_1->Get(weight_index));
+    Tensor& momentums_2 = const_cast<Tensor&>(p.momentums_2->Get(weight_index));
+
+    // Compute exponentially-averaged historical gradient.
+    MakeEigenArrayMap<T>(momentums_1) = alpha_ * MakeEigenArrayMap<T>(momentums_1) +
+                                        (1.f - alpha_) * MakeEigenArrayMap<T>(gradient);
+
+    // Compute exponentially-averaged historical squared gradient.
+    MakeEigenArrayMap<T>(momentums_2) = beta_ * MakeEigenArrayMap<T>(momentums_2) +
+                                        (1.f - beta_) * MakeEigenArrayMap<T>(gradient) * MakeEigenArrayMap<T>(gradient);
+
+    auto denom = MakeEigenArrayMap<T>(momentums_2).sqrt() + epsilon_;
+    MakeEigenArrayMap<T>(weight) = MakeEigenArrayMap<T>(weight) -
+                                   (lr_corrected * MakeEigenArrayMap<T>(momentums_1) / denom);
+
+    // Perform weight decay.
+    MakeEigenArrayMap<T>(weight) = MakeEigenArrayMap<T>(weight) - (lr * weight_decay_ * MakeEigenArrayMap<T>(weight));
+  }
+  return Status::OK();
+}
+
+template <typename T>
+Status AdamWOptimizer<T>::CopyInputTensorToOutputTensor(const Tensor& source_tensor, Tensor& dest_tensor) const {
```

Comment:
Does this need to be defined in the `AdamWOptimizerBase`? Can it not be defined in the unnamed namespace for both cpu and cuda implementations?

Reply:
For CPU, we don't need implement as a function of AdamWOptimizerBase. 
For CUDA, because we need access Stream() to get the current working stream, so we have to do this way.

Reply:
optional: can the stream not be passed in? I think it is fine, but it would be nice if we can move things out of the header file if possible.

Reply:
the stream can be only accessible by calling into cuda kernel's Stream(), so it is not passed in this function as params.